### ISSUE TITLE
img src changes to http://example.org after a partial broadcast_update

### ISSUE THREAD

**Issue by giuseb** [rails/rails#45146]
### Steps to reproduce

The model:
```ruby
class User < ApplicationRecord
  has_one_attached :avatar
end
```

The view:
```erb
<%= turbo_stream_from @user %>
<p><%= @user.name %></p>
<div id="avatar">
  <%= render partial: "users/avatar", locals: {user: @user} %>
</div>
```

The partial:
```erb
<%= image_tag user.avatar %>
```

With this resulting HTML, the image looks fine:
```html
<img src="http://localhost:3000/rails/active_storage/blobs/redirect/YADAYADA/b6dd.jpg">
```

Now, however, in the console:
```irb
irb(main):001:0> User.last.broadcast_update target: "avatar", partial: "users/avatar"
```
### Expected behavior

Nothing should change on the page

### Actual behavior

In the broadcast, `localhost` is replaced with `example.org`:

```
Turbo::StreamsChannel transmitting "<turbo-stream action=\"update\" target=\"avatar\"><template><img src=\"http://example.org/rails/active_storage/blobs/redirect/YADAYADA/b6dd.jpg\" />\n</te... (via streamed from Z2lkOi8vaGd1aWRlL1VzZXJ2aWV3LzE2Mw)
```

The image is now broken, as the rendered HTML changes to:

``` html
<div id="avatar>
  <img src="http://example.org/rails/active_storage/blobs/redirect/YADAYADA/b6dd.jpg">
</div>
```

_**Please note that the same problem occurs in a production environment.**_

### System configuration

- ruby 3.1.2
- rails 7.0.3

**Comment by luanzeba** [comment]
Hi @giuseb , thanks for opening this issue!

* I created a sample application and was able to reproduce the issue from there https://github.com/luanzeba/repro-rails-45146
* After some investigation I noticed the `example.org` value is coming from this hash https://github.com/rails/rails/blob/25b126707152de3a1ec9762ee564f3c7623373b3/actionpack/lib/action_controller/renderer.rb#L40
  * I was able to prove that by changing the value in the hash and seeing the new value come through.
* I was able to work around the issue by setting the following configuration in `config/environments/development.rb`
  * You should be able to configure that in other environments as well, although I didn't test it.
```ruby
 config.action_controller.default_url_options = { host: 'http://localhost', port: 3000 }
```
  
It seems to me that the defaults could be better, but I'm not sure how exactly I'd go about it since you're broadcasting changes from the Rails console

**Comment by giuseb** [comment]
Hello Luan, thanks for the feedback!
Please note that I used the console to demonstrate the problem in the simplest possible way.
However, I first observed the behavior when issuing `broadcast_update` from inside a sidekiq job.
So I guess the question is why that `http_host: "example.org"` applied in some circumstances (turbo stream, background job) but not during a normal request.

**Comment by matthee** [comment]
Hi @giuseb!

I've experienced similar problems in the past. The easiest way to fix your problem is to set `default_url_options` as suggested above. 

To answer your question: Background jobs and the rails console are run outside the request/response cycle of a HTTP request. In these situations Rails has no way of knowing what your current host is and defaults to `example.org`.

**Comment by giuseb** [comment]
Matthias & Luan, thanks for the clarfication, now I understand.
I wonder, though, would it **ever** be appropriate to use `example.org` as a fallback URL?
In other words, wouldn't it be better to raise an error in situations where rails needs the current host but does not know what it is?

Giuseppe

**Comment by manafire** [comment]
In my case I was using `ENV.fetch("PORT", 3000)` to set the `default_url_options` port which was causing it to pick up Cable port instead of the server.  Hard-coding the port fixed the issue for me.

**Comment by giuseb** [comment]
> I don't think that's the best option either. Maybe raising an error would be better.. @giuseb , would you like to attempt a PR with that change? Otherwise I'd like to close out this issue since we have identified the root cause and a workaround.

I would love to help, but I am not sure I have the necessary knowledge to put together a proper PR.

I too had spotted `example.org` inside a `DEFAULTS` hash, in  `rails/actionpack/lib/action_controller/renderer.rb`.

Now, DEFAULTS appears to be used only once in this file:

```ruby
def self.for(controller, env = {}, defaults = DEFAULTS.dup)
  new(controller, env, defaults)
end
```

In turn, `Renderer.for` seems to be called only once, in `rails/actionpack/lib/action_controller/metal/rendering.rb`:

```ruby
def setup_renderer! # :nodoc:
  @renderer = Renderer.for(self)
end
```

And at this point, sorry, I feel utterly lost. Unless I am missing some function-calling pathway, it looks like `Renderer.for` is never passed arguments other than the controller itself. Thus, `defaults` always defaults to `DEFAULTS` 😅...

**Comment by hahmed** [comment]
It _looks_ like this will be fixed via https://github.com/rails/rails/pull/39629 (or the superseded PR linked)

**Comment by giuseb** [comment]
> Hey @giuseb, would you like to try out the fix in this PR and provide any feedback there #42520? It seems like the author is trying to get it merged

I feel bad about this, Luan, but I won't be able to. I have wanted to contribute to the repo for a long time, but so far I have only suggested documentation fixes. I would need a _simpler_ task than this to get started! 😅
Too many moving parts for my limited understanding of rails' innards...

**Comment by luanzeba** [comment]
Hi @giuseb , that's absolutely fair. I also don't fully understand how this feature works, so I'm not able to help much.

That being said, I will close this issue as a duplicate of https://github.com/rails/rails/issues/39566
